cmd/hooks/prestop.go

```diff
@@ -0,0 +1,175 @@
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an 'AS IS' BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hooks
+
+import (
+	"context"
+	"fmt"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"os"
+	"sigs.k8s.io/aws-fsx-csi-driver/pkg/driver"
+)
+
+/*
+When a node is terminated without first unmounting clients, this leaves stranded clients server side, which in turn need to be
+evicted by the Lustre filesystem.
+
+This PreStop lifecycle hook aims to ensure that before the node (and the CSI driver node pod running on it) is shut down,
+there are no more pods with PVCs on the node, thereby indicating that all volumes have been successfully unmounted and detached.
+
+No unnecessary delay is added to the termination workflow, as the PreStop hook logic is only executed when the node is being drained
+(thus preventing delays in termination where the node pod is killed due to a rolling restart, or during driver upgrades, but the workload pods are expected to be running).
+If the PreStop hook hangs during its execution, the driver node pod will be forcefully terminated after terminationGracePeriodSeconds.
+*/
+
+const clusterAutoscalerTaint = "ToBeDeletedByClusterAutoscaler"
+const v1KarpenterTaint = "karpenter.sh/disrupted"
+const v1beta1KarpenterTaint = "karpenter.sh/disruption"
+
+// drainTaints includes taints used by K8s or autoscalers that signify node draining or pod eviction
+var drainTaints = map[string]struct{}{
+	v1.TaintNodeUnschedulable: {}, // Kubernetes common eviction taint (kubectl drain)
+	clusterAutoscalerTaint:    {},
+	v1KarpenterTaint:          {},
+	v1beta1KarpenterTaint:     {},
+}
+
+func PreStop(clientset kubernetes.Interface) error {
+	klog.InfoS("PreStop: executing PreStop lifecycle hook")
+
+	nodeName := os.Getenv("CSI_NODE_NAME")
+	if nodeName == "" {
+		return fmt.Errorf("PreStop: CSI_NODE_NAME missing")
+	}
+
+	node, err := fetchNode(clientset, nodeName)
+	if err != nil {
+		return err
+	}
+
+	if isNodeBeingDrained(node) {
+		klog.InfoS("PreStop: node is being drained, checking for remaining pods with PVCs", "node", nodeName)
+		return waitForPodShutdowns(clientset, nodeName)
+	}
+
+	klog.InfoS("PreStop: node is not being drained, skipping pods check", "node", nodeName)
+	return nil
+}
+
+func fetchNode(clientset kubernetes.Interface, nodeName string) (*v1.Node, error) {
+	node, err := clientset.CoreV1().Nodes().Get(context.Background(), nodeName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("fetchNode: failed to retrieve node information: %w", err)
```

Comment:
again - return err here?

Reply:
same logic as other places

Reply:
This does return an error. The nil is for the node object, which makes sense if you are unable to fetch it.